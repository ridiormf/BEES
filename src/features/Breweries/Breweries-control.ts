import React from 'react';
import { useApplicationContext } from '../../providers/ApplicationProvider';
import { useBreweriesContext } from '../../providers/BreweriesProvider';

export const useBreweriesControl = () => {
  const { closeFullLoading, showNewFeedback } = useApplicationContext();
  const { breweriesResources, breweries, setOrUpdateBreweries } =
    useBreweriesContext();

  const breweriesResponse = breweriesResources?.read();

  React.useEffect(() => {
    if (breweriesResponse?.data && !breweries) {
      setOrUpdateBreweries!(breweriesResponse?.data);
    }
  }, [breweriesResponse?.data, breweries]);

  const onSaveNewBullet = (cardId: string | number, value: string) => {
    const brewery = breweries?.find((item) => `${item.id}` === cardId);
    if (brewery && value) {
      if (!brewery.customBullets) {
        brewery.customBullets = [];
      }
      brewery.customBullets.push(value);
      const newBreweries = breweries?.map((item) =>
        item.id === brewery.id ? brewery : item,
      );
      setOrUpdateBreweries!(newBreweries!);
    }
  };

  const onDeleteBrewery = (cardId: string | number) => {
    const newBreweries = breweries?.filter((item) => {
      return item.id !== cardId;
    });

    setOrUpdateBreweries!(newBreweries!);
  };

  React.useEffect(() => {
    if (breweriesResponse) {
      closeFullLoading!();
      if (!breweriesResponse.ok) {
        showNewFeedback!({
          kind: 'error',
          message: 'something went wrong when trying to query the breweries.',
        });
      }
    }
  }, [breweriesResponse]);

  return {
    state: {
      breweries,
    },
    methods: {
      onSaveNewBullet,
      onDeleteBrewery,
    },
  };
};
