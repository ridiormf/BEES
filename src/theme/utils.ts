type Space = number | string;

const utils = {
  marginHorizontal: (value: Space) => ({
    marginLeft: value,
    marginRight: value,
  }),
  marginVertical: (value: Space) => ({
    marginTop: value,
    marginBottom: value,
  }),
  paddingHorizontal: (value: Space) => ({
    paddingLeft: value,
    paddingRight: value,
  }),
  paddingVertical: (value: Space) => ({
    paddingTop: value,
    paddingBottom: value,
  }),
};

export default utils;
