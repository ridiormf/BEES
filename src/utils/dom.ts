export const getBody = () => {
  return document.getElementsByTagName('body')[0];
};
