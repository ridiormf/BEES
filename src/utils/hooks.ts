import React from 'react';
import { useLocation, useNavigate } from 'react-router-dom';

/**
 * Function that simulates the behavior of a class component constructor,
 * adapted to hooks, for functional components.
 *
 * @param {function} callback Callback which will be called as soon as the component is instantiated.
 */
export const useConstructor = <ValueType = void>(
  callback: () => ValueType,
): ValueType | void => {
  const componentWasBuilt = React.useRef<boolean>(false);
  const value = React.useRef<ValueType>();
  if (!componentWasBuilt.current) {
    componentWasBuilt.current = true;
    value.current = callback();
  }

  return value.current;
};

/**
 * Function that simulates the behavior of a class component method "componentDidMount",
 * adapted to hooks, for functional components.
 *
 * @param {function} callback Callback which will be called as soon as the component mount.
 */
export const useDidMount = (callback: () => void): void => {
  React.useEffect(callback, []);
};

/**
 * Function that simulates the behavior of a class component method "componentWillUnmount",
 * adapted to hooks, for functional components.
 *
 * @param {function} callback Callback which will be called as soon as the component will unmount.
 */
export const useWillUnmount = (callback: () => void): void => {
  React.useEffect(() => {
    return callback;
  }, []);
};

/**
 * Function that simulates the behavior of a class component method "componentDidUpdate",
 * adapted to hooks, for functional components.
 *
 * @param {function} callback Callback which will be called on each value change of variables declared as dependencies,
 * excluding the first render of the component.
 * @param {React.DependencyList} deps Dependency list to be listened.
 */
export const useDidUpdate = (
  callback: () => void,
  deps: React.DependencyList,
): void => {
  const firstRender = React.useRef<boolean>(true);
  React.useEffect(() => {
    if (firstRender.current) {
      firstRender.current = false;
    } else {
      callback();
    }
  }, deps);
};

/**
 * Function that use main hooks from "react-router-dom" to simplify usability.
 *
 * @returns {object} containing following attributes:
 * @type {function} "goBack": to return navigation one page.
 * @type {function} "go": to navigate for path.
 * @type {string} "pathname": current url pathname.
 */
export const useAppNavigation = () => {
  const navigate = useNavigate();
  const location = useLocation();

  const goBack = () => {
    navigate(-1);
  };

  const go = (path: string) => {
    navigate(path);
  };

  return {
    goBack,
    go,
    pathname: location.pathname,
  };
};
