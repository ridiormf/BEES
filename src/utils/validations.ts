/**
 * Function that checks if the full name is valid.
 * It will check:
 * 1. if the string has, at least, two names, and if all names has 2 letters or more;
 * 2. If the string not start or end with blank space;
 * 3. If it doesn't contain numbers; and
 * 4. If it doesn't contain two or more blank spaces between words
 *
 * @param {string} fullName String to be checked.
 * @returns {boolean} Return true if is a valid name, otherwise returns false.
 */
export const checkFullName = (fullName?: string): boolean => {
  if (!fullName) return false;

  const splittedNames = fullName.split(' ');

  return (
    splittedNames.length > 1 &&
    splittedNames.every((name) => name.length > 1 && !/\d/g.test(name))
  );
};
