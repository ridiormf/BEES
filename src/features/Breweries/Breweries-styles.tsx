import { CSS } from '@stitches/react';
import theme from '../../theme';

const ErrorContainer = theme.styled('div', {
  display: 'flex',
  flexGrow: 1,
  alignSelf: 'stretch',
  alignItems: 'center',
  justifyContent: 'center',
  flexDirection: 'column',
});

const ErrorTitle = theme.styled('h2', {
  fontSize: 30,
  fontWeight: 400,
});

const ImageButton = theme.styled('button', {
  padding: 0,
  lineHeight: 0,
  border: 'none',
  backgroundColor: 'unset',
});

const imageCSS: CSS = {
  filter: 'grayscale(50%)',
  opacity: 0.7,
  transition: '200ms',
  cursor: 'pointer',
  '&:hover': {
    filter: 'grayscale(0%)',
    opacity: 1,
  },
};

export const breweriesStyles = {
  styled: {
    ErrorContainer,
    ErrorTitle,
    ImageButton,
  },
  css: {
    image: imageCSS,
  },
};
