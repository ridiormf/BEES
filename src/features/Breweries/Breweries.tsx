import React from 'react';
import { CardListDataItem } from '../../components/CardList';
import { useBreweriesControl } from './Breweries-control';
import { breweriesStyles } from './Breweries-styles';

import { CardListBulletProps } from '../../components/CardList/CardList-types';

import chartSvg from '../../assets/svg/chart-square-bar.svg';
import pinSvg from '../../assets/svg/location-marker.svg';
import phoneSvg from '../../assets/svg/phone.svg';
import beesLogo from '../../assets/png/bees.png';
import Image from '../../components/Image';

const CardList = React.lazy(() => import('../../components/CardList'));

const {
  styled: { ErrorContainer, ErrorTitle, ImageButton },
  css,
} = breweriesStyles;

const Breweries = () => {
  const {
    state: { breweries },
    methods: { onSaveNewBullet, onDeleteBrewery, tryQueryAgain },
  } = useBreweriesControl();

  return breweries?.map ? (
    <CardList
      bulletsWithAddMore
      onSaveNewBullet={onSaveNewBullet}
      onDeleteCard={onDeleteBrewery}
      data={breweries.map((brewery): CardListDataItem => {
        const bullets: Array<CardListBulletProps> = [];
        if (brewery.brewery_type) {
          bullets.push({
            icon: chartSvg,
            children: brewery.brewery_type,
          });
        }
        if (brewery.postal_code) {
          bullets.push({
            icon: pinSvg,
            children: brewery.postal_code,
          });
        }
        if (brewery.phone) {
          bullets.push({
            icon: phoneSvg,
            children: brewery.phone,
          });
        }
        if (brewery.customBullets) {
          bullets.push(
            ...brewery.customBullets.map((value) => ({
              children: value,
            })),
          );
        }
        return {
          key: brewery.id,
          title: brewery.name,
          firstLineText: brewery.street,
          firstLineEmptyText: 'No street registered',
          secondLineText:
            brewery.city && brewery.state && brewery.country
              ? `${brewery.city}, ${brewery.state} - ${brewery.country}`
              : '',
          secondLineEmptyText: 'Location not registered or incomplete',
          bullets,
        };
      })}
    />
  ) : (
    <ErrorContainer>
      <ErrorTitle>No brewery to show</ErrorTitle>
      <ImageButton onClick={tryQueryAgain}>
        <Image src={beesLogo} css={css.image} />
      </ImageButton>
      <ErrorTitle>Tap the BEE to query again.</ErrorTitle>
    </ErrorContainer>
  );
};

export default Breweries;
