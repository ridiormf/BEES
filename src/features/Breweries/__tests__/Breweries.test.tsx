import { render } from '@testing-library/react';
import { Wrapper } from '../../../__tests__/utils';
import Breweries from '../Breweries';

describe('features ->', () => {
  describe('Breweries ->', () => {
    describe('snapshots ->', () => {
      it('Should match snapshot', () => {
        const { container } = render(<Breweries />, {
          wrapper: Wrapper,
        });
        expect(container).toMatchSnapshot();
      });
    });
  });
});
