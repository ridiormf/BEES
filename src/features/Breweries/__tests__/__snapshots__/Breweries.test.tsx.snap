// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components -> Breweries -> snapshots -> Should match snapshot 1`] = `
<div>
  <div
    class="c-fgfIsx"
  >
    <h2
      class="c-biZjsf"
    >
      No brewery to show
    </h2>
    <button
      class="c-eJdPgB"
    >
      <img
        alt="Bee image"
        class="PJLV PJLV-ijfrJSw-css"
        src="bees.png"
      />
    </button>
    <h2
      class="c-biZjsf"
    >
      Tap the BEE to query again.
    </h2>
  </div>
</div>
`;

exports[`features -> Breweries -> snapshots -> Should match snapshot 1`] = `
<div>
  <div
    class="c-fgfIsx"
  >
    <h2
      class="c-biZjsf"
    >
      No brewery to show
    </h2>
    <button
      class="c-eJdPgB"
    >
      <img
        alt="Bee image"
        class="PJLV PJLV-ijfrJSw-css"
        src="bees.png"
      />
    </button>
    <h2
      class="c-biZjsf"
    >
      Tap the BEE to query again.
    </h2>
  </div>
</div>
`;
